### ISSUE TITLE
v3.5.0 breaks reactive render

### ISSUE THREAD

**Issue by jeffgran** [mobxjs/mobx-react#85]
Had a production bug come in yesterday that "the page doesn't update when I click x button".

I finally tracked down the change that broke our functionality:

```javascript
                            if (self.__$mobxMounted)
                                React.Component.prototype.forceUpdate.call(self)
```

If I delete the `if` conditional and always update, my code works again. 

I know this isn't a very good bug report, with no specific means to reproduce. But wanted to get a discussion started. I have reverted to 3.4.0 in production for now until I can figure out what to do.

**Comment by mweststrate** [comment]
@jeffgran That's nasty! Apologies, breaking stuff is never the intention, so try to maintain an extensive suite with exotic edge cases. If you would be able to pin point yours that would be awesome!

Which react version are you using? Did you by any chance see this warning passing by in the logs of the working version?
```
Warning: forceUpdate(...): Cannot update during an existing state transition (such as within `render` or another component's constructor). Render methods should be a pure function of props and state; constructor side-effects are an anti-pattern, but can be moved to `componentWillMount`.
```

**Comment by jeffgran** [comment]
No, I don't get that error in the working case.

I am trying to reproduce in a minimal case, but no luck yet. 

In my real code, I am able to switch back and forth and see it working/not working. It seems like it has to do with the `this.render` and `initialRender`, which runs once, and then assigns `reactiveRender` to `this.render`. So `initalRender` only gets called once. If, in that initial (only) call, it calls `forceUpdate`, then everything works. If it doesn't, then it doesn't. I don't know what is happening inside `forceUpdate` but it seems to trigger the reactive render into working.

I guess that suggests a race condition of some kind. In my code, during `initialRender`, `this.__$mobxMounted` is undefined, so the `forceUpdate` does not get called, and then after that it never gets called.

One weird trick to make this work is to call `React.Component.prototype.forceUpdate.call(this);` inside `componentDidMount`. Making JUST that change, it works correctly.

What could be happening inside `forceUpdate` that changes the behavior?

**Comment by jeffgran** [comment]
Also, forgot to answer you: react version 0.14.8 -- 90% sure I tried with 15.x yesterday and that didn't work either.

**Comment by mweststrate** [comment]
@jeffgran could you test whether either `mobx-react@3.5.0-fix85`  or `mobx-react@3.5.0-fix85-2` fixes the issue?

**Comment by jeffgran** [comment]
Yes! Commit[0d9f567](https://github.com/mobxjs/mobx-react/commit/0d9f5670ceb4a6ed769c2985895f27c41385ca2e) fixes it for me.

I guess it was caused by something in my constructor triggering an update before the thing was even mounted?

**Comment by mweststrate** [comment]
yep i think so.  but that is hard to find in a real project. did the second
version also fix it? that has a nice patch but didn't commit it yet. will
release definitive fix tomorrow

Op wo 13 jul. 2016 22:44 schreef jeffgran <notifications@github.com>:

**Comment by jeffgran** [comment]
Yes, v3.5.0-fix85-2 does work as well.

Thanks again for all your work on mobx and for the quick response on this. Very much appreciated.

**Comment by mweststrate** [comment]
Just released `3.5.1`. That should fix the issue!

**Comment by jeffgran** [comment]
👍
